### PR TITLE
[Issue #6079] Enable spatial data stores for Django database

### LINKED ISSUE #4346: Search by extent broken
I get zero results from API if I search by extent. no matter how much I zoom out on the layers search page.

Current master, spc.

### PR DESCRIPTION
**Pull request by mtnorthcott**
Closes #4346

This PR is now a major component of [GNIP-75](https://github.com/GeoNode/geonode/issues/6079) and enables spatial Django datastores for GeoNode.

## Previous Description
This PR aims to enable full functionality of the 'search by extent' feature when filtering layers.

The previous implementation stored bounding box coordinates as decimal fields in the database, performing complex custom methods to infer coordinates and search queries such as filtering by layer extent.

To address issues in #4346, this PR:
* Uses the PostGIS extension in the GeoNode database (as specified by `DATABASE_URL=` in `geonode/settings.py`)
* Replaces existing decimal-based bounding fields with a PostGIS polygon (includes data migration)
* Relies on PostGIS to perform any queries related to bounding boxes

As a side effect, bounding boxes not provided in EPSG4326 can now be reprojected on-the-fly.

Previous decimal `bbox_` fields have not been removed yet, but such a migration can be added once people are happy with our implementation.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: flake8 geonode
- [x] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Comment by afabiani** [comment]
@mtnorthcott thanks for the effort, for sure this code could be the base for future improvements. Nevertheless, the PR cannot be merged right now for several reasons:

1. This forces people to use postgis by dropping the default sqlite DB, which is not bad per see, but such a huge change in the default behavior requires a GNIP. For sure it cannot be included in the next release.

2. Big changes on Travis too must be done in order to pass the tests; other than this I don't know how it behaves with all the GIS clients relying on the old bbox_* properties of the layers

3. Least but not last, I already did a [commit ](https://github.com/GeoNode/geonode/commit/66247908448f964139a3b96b5cc2b659cf382e23#diff-01d93696899f392996ab0326a506a0e1) in order to "mitigate" the issue. I'm aware this is not the optimal solution, but it works with minimal changes

Long story short, I vote to postpone this after the release and open a thread and discussion on an appropriate GNIP.

**Comment by GeoTob** [comment]
Thanks @afabiani - I think @mtnorthcott has followed exactly the steps outlined on https://github.com/GeoNode/geonode/issues/4346#issuecomment-583636963 and then added into 3.0 on Feb 11 and Apr 3. Bit surprised to see the comment at this stage to be honest. 

That said, I agree with your point - switching to PostGIS / Spatialite requires proper process. @afabiani would you be able to give @mtnorthcott some pointers and assistance through the GNIP process? E.g. docs say mailing list, but I see many GNIPs discussed as repo issues? 

As an aside, things like `GIS clients relying on the old bbox_* properties` should be very easy to solve in a pythonic / Django way. I'm very keen to see complex custom code removed going forward :)

**Comment by afabiani** [comment]
I'm not saying this not a good solution. I'm just saying we're far to be able to accept it. Especially we need to be sure the documentation won't be misleading for people which want to install geonode or update it.

**Comment by gannebamm** [comment]
I do understand your view and thoughts @afabiani. IIRC you were not possible to attend the video meeting which @GeoTob is talking about. In there we talked about the GeoNode 3.0 issues and @t-book later sent this concept to the dev-list and/or psc-list to comment on it.
@afabiani is right: This will change parts deep inside of GeoNode and would demand users to use a spatial backend. This will maybe cause a breaking change for some of them. But, how many users do use GeoNode without _a proper_ spatial backend? As I already said many times I am very in favor of pinning GeoNode to particular software solutions and make it slimmer instead of ORMing our way in every direction. But this needs to be discussed and voted by the PSC.

So long story short: A GNIP has to be created (which is an issue template https://github.com/GeoNode/geonode/issues/new/choose). Please keep it slim and refer to this PR and its issue. There is already loads of documentation about your technical concept and discussion as well as an alternative workaround by Alessio. Please refer to those and link them. The PSC will than discuss and vote about the release process.

**Comment by afabiani** [comment]
@gannebamm you are right, IMHO before merging a PR like this we must be sure at least that:

1. All the documentation, especially the setup parts, is updated accordingly.
2. All the possible ways to install GeoNode, including Dockers, have been updated and tested
3. Travis and tests are all updated accordingly
4. New changes are well advertised

For the time being those are the main three points preventing this to be merging.
My opinion would be to collect all those points and checks on a dedicated GNIP.

**Comment by t-book** [comment]
@afabiani @gannebamm thanks for sharing your thoughts. I agree with pretty much all of your points. Removing complex custom code as @GeoTob nicely said is very welcome as well. On the other hand, I'm a bit afraid of introducing things into the codebase which could stop paver setup from being fully functional.

> My opinion would be to collect all those points and checks on a dedicated GNIP.

+1 // I am looking forward to reading other opinions

**Comment by GeoTob** [comment]
@mtnorthcott 

> You should open an new issue and choose GNIP from the issue template
> Just follow the template questions. Should be self explaining. When ready. Just drop a line at the mailinglist and asking for review. One line is enough and all PSC members should be aware of the new issues.
> @t-book will link your GNIP in the Wiki.

As others say above, please keep it short & slim but try to identify potential issues and how you propose to resolve them as part of the MR. That's for code and docs re GN-paver, GN-docker, upgrade paths and ideally downgrade paths.

File: .travis.yml [hunk 1]
```diff
@@ -308,14 +308,15 @@ install:
       wget https://github.com/mozilla/geckodriver/releases/download/$GECKODRIVER_VERSION/geckodriver-$GECKODRIVER_VERSION-linux64.tar.gz -O geckodriver.tar.gz;
       mkdir bin;
       tar zxf geckodriver.tar.gz -C bin;
+      git clone -b search-by-extent-postgis https://github.com/mtnorthcott/postgis-docker.git;
```
Comment (afabiani) [inline .travis.yml:311]: Is this still necessary?

Comment (mtnorthcott) [reply .travis.yml:311]: Not necessary once the change in https://github.com/GeoNode/postgis-docker/pull/2 is available in the Docker hub.

Comment (afabiani) [reply .travis.yml:311]: @mtnorthcott pr mrged and docker image rebuilt, can we remove this one please?

**Comment by GeoTob** [comment]
@mtnorthcott @afabiani would it be useful for you two to have a chat some time soon and discuss what remains to be done? 

I think @mtnorthcott will have some time next week to bring this branch up to date with master again and address any remaining tasks.

**Comment by mtnorthcott** [comment]
I haven't been successful in resolving the failing Travis CI build so far. The issue is that a certain SQL query made by calling `MapSnapshot.objects.create(...)`/`Map.objects.create(...)`/`Map(...)` will cause a segmentation fault, preventing the rest of the tests from running. The tests in question are those in `geonode.maps.tests.MapsTest`. Logging queries issued in `django/db/backends/sqlite3/base.py` yields the following output:
```sql
...
SAVEPOINT "s140146508863232_x1334";
RELEASE SAVEPOINT "s140146508863232_x1334";
SAVEPOINT "s140146508863232_x1335";
RELEASE SAVEPOINT "s140146508863232_x1335";
SAVEPOINT "s140146508863232_x1336";
RELEASE SAVEPOINT "s140146508863232_x1336";
ROLLBACK TO SAVEPOINT "s140146508863232_x1273";
```

I have tried:
* Upgrading the Travis CI runner to operate using `dist: bionic` so that a more recent SQLite version is installed
* Rolling back to https://github.com/GeoNode/geonode/pull/6064/commits/4e80b6c39259e76f122306e4f7f10327caf9e4cf where I made [a comment saying that Travis CI tests were working](https://github.com/GeoNode/geonode/issues/6079#issuecomment-639143053)
* Using an on-disk database for Django and geodatabases
* Removing cache for the job

I cannot reproduce this outcome locally using `paver run-tests`. However, mounting the base directory as a volume of a xenial-based Docker container, installing prerequisites and running the same command does work.

@afabiani Do you have any ideas or advice on what I could do next to help resolve the problem?

File: geonode/client/templates/ol/layers/layer_ol2_map.html [hunk 2]
```diff
@@ -13,7 +13,7 @@
 <script type="text/javascript">
   document.addEventListener("DOMContentLoaded", function (event) {
     {% if resource.ll_bbox %}
-      var bbox = [{{ resource.ll_bbox_string }}];
+      var bbox = [{{ resource.bbox_string }}];
```
Comment (afabiani) [inline geonode/client/templates/ol/layers/layer_ol2_map.html:16]: The bbox here should always be returned into 4326 SRS. Please, revert to ll_bbox_string.

File: geonode/templates/_permissions_form.html [hunk 3]
```diff
@@ -341,7 +341,7 @@ <h4 class="modal-title" id="exampleModalLabel">{% trans "Message box" %}</h4>
 
     /* OpenLayers */
     {% if resource.ll_bbox %}
-      var bbox = [{{ resource.ll_bbox_string }}];
+      var bbox = [{{ resource.bbox_string }}];
```
Comment (afabiani) [inline geonode/templates/_permissions_form.html:344]: The bbox here should always be returned into 4326 SRS. Please, revert to ll_bbox_string.

File: geonode/serializer.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+from functools import partial
```
Comment (afabiani) [inline geonode/serializer.py:1]: New files must contain headers.

File: geonode/base/models.py [hunk 5]
```diff
@@ -968,54 +952,61 @@ def group_name(self):
...
-            llbbox[3],  # y1
-            self.srid]
+        """BBOX is in the format [x0, x1, y0, y1, "EPSG:srid"]. Provides backwards
+        compatibility after transition to polygons."""
+        bbox = self.bbox_polygon
+        if bbox.srid != 4326:
+            bbox = bbox.transform(4326, clone=True)
 
-    @property
-    def ll_bbox_string(self):
```
Comment (afabiani) [inline geonode/base/models.py:998]: Please restore this property. This might be used by client libraries to retrieve the lon-lat bbox representation.
Removing this might break backward compatibility.

**Review by afabiani — CHANGES_REQUESTED** [review]
@mtnorthcott I've added few comments. Please take a look.

Moreover, there's a Travis test failing

```
======================================================================
FAIL: test_geonode_rest_layer_uploader (geonode.layers.tests.LayersUploaderTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/GeoNode/geonode/geonode/decorators.py", line 57, in wrapper
    return func(*args, **kwargs)
  File "/home/travis/virtualenv/python3.7.6/lib/python3.7/site-packages/django/test/utils.py", line 373, in inner
    return func(*args, **kwargs)
  File "/home/travis/build/GeoNode/geonode/geonode/layers/tests.py", line 1341, in test_geonode_rest_layer_uploader
    self.assertEqual(data['bbox'], _l.bbox_string)
AssertionError: '99.165148,121.54266,21.5124,39.93143000000001' != '99.165148,121.54266,21.512399999999996,39.931430000000006'
- 99.165148,121.54266,21.5124,39.93143000000001
?                           ^                 ^
+ 99.165148,121.54266,21.512399999999996,39.931430000000006
?                           ^^^^^^^^^^^^                 ^^
----------------------------------------------------------------------
```

**Comment by afabiani** [comment]
@mtnorthcott we are almost there!!

The PR has still some issues that I'm going to report here below.

I haven't found any major regression other than those ones for the time being. Will test again deeply once those issues will be sorted out.

## Set Thumbnail method seems to not correctly compute the bbox

Trying to set a new thumbnail from GeoNode automatically, always results in a gray tile (see the figure below)

image

## Save Map gives an error

This is something related to the mapstore-adapter; I will take care of this

```
Traceback (most recent call last):
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/mapstore2_adapter/plugins/serializers.py", line 259, in set_geonode_map
    map_obj = Map(
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/modeltranslation/translator.py", line 264, in new_init
    old_init(self, *args, **kwargs)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/modeltranslation/translator.py", line 264, in new_init
    old_init(self, *args, **kwargs)
  File "/opt/geonode/geonode/base/models.py", line 854, in __init__
    super(ResourceBase, self).__init__(*args, **kwargs)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/polymorphic/models.py", line 198, in __init__
    super(PolymorphicModel, self).__init__(*args, **kwargs)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/db/models/base.py", line 501, in __init__
    raise TypeError("%s() got an unexpected keyword argument '%s'" % (cls.__name__, kwarg))
TypeError: Map() got an unexpected keyword argument 'bbox_x0'
```
## Remote Service Layer Harvesting produces an error

```
Remote Services - Import Resources

Layer() got an unexpected keyword argument 'bbox_x0'
Remote Layer [rt_sfondo.batimetriche@150dpi] couldn't be updated
```

## Migrations Error

```
Operations to perform:
  Apply all migrations: account, actstream, admin, announcements, auth, avatar, base, br, contenttypes, dialogos, documents, favorite, geonode_client, geonode_themes, groups, guardian, invitations, layers, maps, mapstore2_adapter, monitoring, oauth2_provider, people, pinax_notifications, ratings, services, sessions, sites, socialaccount, taggit, tastypie, upload, user_messages
Running migrations:
  Applying base.0044_resourcebase_bbox_polygon... OK
  Applying base.0045_auto_20200507_0445... OK
  Applying base.0046_auto_20200929_0811... OK
  Applying geonode_themes.0011_auto_20200727_1451... OK
  Applying layers.0028_auto_20200323_1945...Traceback (most recent call last):
  File "manage.py", line 29, in <module>
    execute_from_command_line(sys.argv)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 232, in handle
    post_migrate_state = executor.migrate(
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/db/migrations/executor.py", line 245, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/afabiani/.virtualenvs/geonode/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/opt/geonode/geonode/layers/migrations/0028_auto_20200323_1945.py", line 10, in populate_polygon
    bbox_tup = (layer.bbox_x0, layer.bbox_y0, layer.bbox_x1, layer.bbox_y1)
AttributeError: 'Layer' object has no attribute 'bbox_x0'
```

This could be probably fixed by adding a simple check like the following one

```python
def populate_polygon(apps, schema_editor):
    Layer = apps.get_model('layers', 'Layer')
    for layer in Layer.objects.all():
        if hasattr(layer, 'bbox_x0'):
            bbox_tup = (layer.bbox_x0, layer.bbox_y0, layer.bbox_x1, layer.bbox_y1)
            layer.bbox_polygon = Polygon.from_bbox(bbox_tup)
            layer.save()
```

**Comment by mtnorthcott** [comment]
@afabiani Errors pertaining to missing or unexpected `bbox_*` attributes should now be fixed as well as the residual Travis test failure. I fixed the former issue by modifying the constructor to `ResourceBase` so that legacy fields are converted to a polygon type before further processing so updates to mapstore_adapater aren't immediately required. However, I was not able to reproduce the problem you were having with automatic thumbnail generation. This works with a clean paver installation method in my environment. Could you please advice?

**Comment by afabiani** [comment]
@mtnorthcott yes, I'll try again. I'll try to keep an eye to other issues too, remote services and map saving.

**Comment by gannebamm** [comment]
Just wanted to drop in and thank @mtnorthcott for the outstanding work!
